[Clang][AArch64] Include SME attributes in the name mangling of function types

kmclaughlin-arm opened the pull request:
Similar to arm_sve_vector_bits, the mangling of function types is implemented as a pseudo template if there are any SME attributes present, i.e.

  `__SME_ATTRS<normal_function_type, sme_state>`

For example, the following function:

  `void f(svint8_t (*fn)() __arm_streaming) { fn(); }`

would be mangled as:

  `_Z1fP11__SME_ATTRSIFu10__SVInt8_tELj1EE`

See https://github.com/ARM-software/acle/pull/358

kmclaughlin-arm committed:
[Clang][AArch64] Include SME attributes in the name mangling of function types.

Similar to arm_sve_vector_bits, the mangling of function types is implemented as
a pseudo template if there are any SME attributes present, i.e.

  __SME_ATTRS<normal_function_type, streaming_mode, za_state, zt0_state>

For example, the following function:

  void f(svint8_t (*fn)() __arm_streaming) { fn(); }

is mangled as:

  fP9__SME_ATTRSIFu10__SVInt8_tELj1ELj0ELj0EE

See https://github.com/ARM-software/abi-aa/pull/290

efriedma-quic commented on clang/lib/AST/ItaniumMangle.cpp:
```diff
@@ -3535,6 +3536,39 @@ void CXXNameMangler::mangleExtFunctionInfo(const FunctionType *T) {
+  Out << (hasSharedState(FunctionType::getArmZAState(SMEAttrs)) ? "Lj1E"
+                                                                : "Lj0E");
+  Out << (hasSharedState(FunctionType::getArmZT0State(SMEAttrs)) ? "Lj1E"
+                                                                 : "Lj0E");
```
Sema currently accepts the following:

```
__arm_new("zt0") void fn_zt0_in(int (*foo)() __arm_in("zt0")) { foo(); }
__arm_new("zt0") void fn_zt0_in(int (*foo)() __arm_inout("zt0")) { foo(); }
```

So either we can't unify the manglings like this, or Sema needs to reject this. (I think the only reasonable way to get Sema to reject is to make the canonical FunctionType indistinguishable.  Which seems a little weird.)

kmclaughlin-arm replied:
Hi @efriedma-quic, these should now receive different mangled names after the changes to https://github.com/ARM-software/acle/pull/358, as all of attributes are represented individually.

kmclaughlin-arm committed:
- Attributes are now encoded as one bitmask, with each of the attributes
  represented separately.
- Removed hasSharedState.

sdesmalen-arm commented on clang/lib/AST/ItaniumMangle.cpp:
```diff
@@ -3536,35 +3536,64 @@ void CXXNameMangler::mangleExtFunctionInfo(const FunctionType *T) {
   }
 }
 
+// The mangling scheme for function types which have SME attributes is implemented as
```
Could you add a link to the (corresponding section in the) AAPCS for this?

sdesmalen-arm commented on clang/lib/AST/ItaniumMangle.cpp:
```diff
@@ -3536,35 +3536,64 @@ void CXXNameMangler::mangleExtFunctionInfo(const FunctionType *T) {
+    return 4;
   default:
-    return false;
+    return 0;
```
I think only `case ARM_None` should return 0. Any other value (default) should be an `llvm_unreachable`.

kmclaughlin-arm committed:
- Added SMEState enum
- Renamed getZAState -> encodeZAState
- Added llvm_unreachable to encodeZAState
- Added a link to the AAPCS in the comment above mangleSMEAttrs

kmclaughlin-arm commented on clang/lib/AST/ItaniumMangle.cpp:
```diff
@@ -3536,35 +3536,64 @@ void CXXNameMangler::mangleExtFunctionInfo(const FunctionType *T) {
+  unsigned Bitmask = 0;
   if (SMEAttrs & FunctionType::SME_PStateSMEnabledMask)
-    Out << "Lj1E";
+    Bitmask |= 1;
```
I've removed details of what each bit represents from the comment now that there is an enum with this information. I've still left in a short description of the template, similar to what has been written for the `arm_sve_vector_bits` attribute.

kmclaughlin-arm committed:
- Changed name of SMEState & changed to enum class
- Renamed encodeZAState -> encodeAAPCSZAState
- Removed Normal from SMEState enum
- Added a link to the relevant section of the AArch64 ACLE doc

sdesmalen-arm commented on clang/lib/AST/ItaniumMangle.cpp:
```diff
@@ -3535,6 +3536,70 @@ void CXXNameMangler::mangleExtFunctionInfo(const FunctionType *T) {
+static unsigned encodeAAPCSZAState(unsigned SMEAttrs) {
+  switch (SMEAttrs) {
+  case FunctionType::ARM_None:
+    return static_cast<unsigned>(AAPCSBitmaskSME::NoState);
```
You can remove the `static_cast` if you use BitmaskEnum, see `LLVM_MARK_AS_BITMASK_ENUM` or  `LLVM_DECLARE_ENUM_AS_BITMASK`.

Unfortunately I found that `operator<<` is not supported (needed below in mangleSMEAttrs), so I've created #118007.

kmclaughlin-arm committed:
- Added LLVM_MARK_AS_BITMASK_ENUM & removed static_casts from encodeAAPCSZAState

kmclaughlin-arm committed:
- Changed LargestValue for LLVM_MARK_AS_BITMASK_ENUM and removed more static_casts
